### PR TITLE
Added session expiry and inactivity expiry

### PR DESCRIPTION
**Pull request by j-maynard**
### What
Implement both the idle and absolute timeouts.
- The idle timeout should be 15 minutes (the session should be invalidated if the last activity was 15 minutes or more ago)
- The absolute timeout should be 90 minutes (the user needs to re-authenticate after 90 minutes of their initial login, regardless of their activity)
The values above should be configurable.

### Why
To comply with security standards and recommendations we should implement session timeouts - both idle and absolute timeouts. This will prevent potential hijackers from using existing sessions or gaining unauthorized access.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -8,9 +8,16 @@ class ApplicationController < ActionController::Base
   before_action :set_user
   before_action :configure_permitted_parameters, if: :devise_controller?
   before_action :check_authorization
+  before_action :check_max_session_time
 
 protected
 
+  def check_max_session_time
+    return false if current_user.nil?
+    time_out = Rails.application.secrets.session_expiry_in_minutes.nil? ? 60 : Rails.application.secrets.session_expiry_in_minutes.to_i
```
Comment (jakubmiarka) [inline app/controllers/application_controller.rb]: it doesn't need to be in secrets, can we define that as a config value and put the default there, instead of here?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -8,9 +8,16 @@ class ApplicationController < ActionController::Base
...
+    sign_out_user if Time.parse(current_user.current_sign_in_at) + time_out.minutes < Time.now
```
Comment (jakubmiarka) [inline app/controllers/application_controller.rb]: being picky, but can we say `timeout` instead of `time_out` and `session_start_time` instead of `current_sign_in_at`?

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -26,4 +33,13 @@ def check_authorization
     flash[:warn] = t('shared.errors.authorisation')
     redirect_to root_path
   end
+
+  private
+
+  def sign_out_user
+    UserSignOutEvent.create(user_id: warden.user.user_id)
```
Comment (jakubmiarka) [inline app/controllers/application_controller.rb]: could we have a new TimeOut event? so we can see how often people timeout?

File: app/models/remote_authenticatable.rb [hunk 4]
```diff
@@ -149,7 +153,9 @@ def serialize_into_session(record)
               user_id: record.user_id,
               roles: record.roles,
               given_name: record.given_name,
-              family_name: record.family_name
+              family_name: record.family_name,
+              current_sign_in_at: record.current_sign_in_at,
```
Comment (jakubmiarka) [inline app/models/remote_authenticatable.rb]: could we parse it to Time object here instead of the controller?

File: app/models/user.rb [hunk 5]
```diff
@@ -26,4 +27,8 @@ def save!; end
   def attributes
     super.merge(roles: nil, permissions: nil)
   end
+
+  def to_hash
+    instance_variables.map { |var| [var.to_s.delete('@'), instance_variable_get(var)] }.to_h
```
Comment (jakubmiarka) [inline app/models/user.rb]: is this related to the PR?

File: app/models/remote_authenticatable.rb [hunk 6]
```diff
@@ -87,6 +87,8 @@ def auth_complete(resp)
         self.given_name = user_attributes['given_name']
         self.family_name = user_attributes['family_name']
         self.mfa = aws_user.preferred_mfa_setting
+        self.current_sign_in_at = Time.now
+        self.last_sign_in_at = Time.now
```
Comment (jakubmiarka) [inline app/models/remote_authenticatable.rb]: do we need this value? isn't devise taking care of that?

File: config/initializers/devise.rb [hunk 7]
```diff
@@ -183,6 +183,8 @@
   # The time you want to timeout the user session without activity. After this
   # time the user will be asked for credentials again. Default is 30 minutes.
   # config.timeout_in = 30.minutes
+  inactivity_expiry = Rails.application.secrets.session_inactivity_in_minutes.nil? ? 15 : Rails.application.secrets.session_inactivity_in_minutes.to_i
+  config.timeout_in = inactivity_expiry.minutes
```
Comment (jakubmiarka) [inline config/initializers/devise.rb]: same like above, set the default in the config instead of here... then this line becomes something like
`config.timeout_in = Rails.application.config.inactivity_timeout`

**Comment by Rosa-Fox** [comment]
We need to remove 'true' from the end of the flash message hehe

flash-message

**Comment by jakubmiarka** [comment]
> We need to remove 'true' from the end of the flash message hehe
> 
> flash-message

true :trollface:

File: spec/system/sign_out_spec.rb [hunk 8]
```diff
@@ -11,4 +11,78 @@
...
+
+    user = FactoryBot.create(:user)
+    sign_in(user.email, user.password)
+    travel_to Time.now + 14.minutes
+    visit(profile_path)
+    expect(current_path).to eql profile_path
+    travel_to Time.now + 16.minutes
+    visit(profile_path)
+    expect(current_path).to eql new_user_session_path
+    expect(page).to have_content 'Your session expired. Please sign in again to continue.'
```
Comment (brenetic) [inline spec/system/sign_out_spec.rb]: We can use the locales here so the test don't break if the text changes. Same for the other few instances above

File: config/environments/development.rb [hunk 9]
```diff
@@ -14,11 +14,17 @@
 
   # Enable/disable caching. By default caching is disabled.
   # Run rails dev:cache to toggle caching.
-  
-  
+
+  # User will be timed out after 90 minutes regardless of activity
+  config.session_expiry_in_minutes = 90
```
Comment (jakubmiarka) [inline config/environments/development.rb]: this can live in `config/application.rb` and we only add this to the `/config/environments/<env>.rb` when we need to overwrite the default... and that might be enough (i.e. no need for env variables)...

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (j-maynard) [reply app/controllers/application_controller.rb]: Timeout event created

File: app/models/user.rb (hunk 5 above)
Comment (j-maynard) [reply app/models/user.rb]: Its been used for debugging a lot both this PR and the previous one.  Its useful to get a hash of the user as apposed to implementing a to_s method.  So I'd like to leave this in for now and flag it for removal once we're doing doing anything user related.

Comment (jakubmiarka) [reply app/models/user.rb]: ok, yeah, that's fine. I just didn't see the correlation with the timeout stuff

File: config/application.rb [hunk 10]
```diff
@@ -35,5 +35,10 @@ class Application < Rails::Application
 
     config.i18n.default_locale = :en
 
+    # User will be timed out after 90 minutes regardless of activity
+    config.session_expiry = 90.minutes
```
Comment (jakubmiarka) [inline config/application.rb]: 👍
